components/ai_chat/core/browser/engine/engine_consumer_conversation_api.cc

```diff
@@ -183,4 +253,104 @@ EngineConsumerConversationAPI::GetAssociatedContentConversationEvent(
+    GetSuggestedTopicsCallback callback) {
+  if (!topics_result.has_value() || topics_result->empty()) {
+    std::move(callback).Run(topics_result);
+    return;
+  }
+
+  base::Value::List topic_list;
+  for (const auto& topic : *topics_result) {
+    topic_list.Append(topic);
+  }
+  std::vector<ConversationEvent> conversation;
+  conversation.push_back({mojom::CharacterType::HUMAN,
+                          ConversationEventType::DedupeTopics,
+                          base::WriteJson(topic_list).value_or(std::string())});
+  api_->PerformRequest(
+      std::move(conversation), "" /* selected_language */,
+      base::NullCallback() /* data_received_callback */,
+      base::BindOnce(
+          [](GetSuggestedTopicsCallback callback,
+             EngineConsumer::GenerationResult result) {
+            // Return deduped topics from the response.
+            std::vector<EngineConsumer::GenerationResult> results = {result};
+            std::move(callback).Run(
+                EngineConsumerConversationAPI::
+                    GetStrArrFromTabOrganizationResponses(results));
+          },
+          std::move(callback)));
+}
+
+void EngineConsumerConversationAPI::ProcessTabChunks(
+    const std::vector<Tab>& tabs,
+    ConversationEventType event_type,
+    base::OnceCallback<void(std::vector<GenerationResult>)> merge_callback,
+    const std::string& topic) {
+  CHECK(event_type == ConversationEventType::GetSuggestedTopicsForFocusTabs ||
+        event_type == ConversationEventType::GetFocusTabsForTopic);
+
+  // Split tab into chunks of 75
+  size_t num_chunks = (tabs.size() + kChunkSize - 1) / kChunkSize;
+  const auto barrier_callback = base::BarrierCallback<GenerationResult>(
+      num_chunks, std::move(merge_callback));
+
+  for (size_t chunk = 0; chunk < num_chunks; ++chunk) {
+    base::Value::List tab_value_list;
+    for (size_t i = chunk * kChunkSize;
+         i < std::min((chunk + 1) * kChunkSize, tabs.size()); ++i) {
+      tab_value_list.Append(base::Value::Dict()
+                                .Set("id", tabs[i].id)
+                                .Set("title", tabs[i].title)
+                                .Set("url", tabs[i].origin.Serialize()));
+    }
+
+    std::vector<ConversationEvent> conversation;
+    conversation.push_back(
+        {mojom::CharacterType::HUMAN, event_type,
+         base::WriteJson(tab_value_list).value_or(std::string()), topic});
+
+    api_->PerformRequest(std::move(conversation), "" /* selected_language */,
+                         base::NullCallback() /* data_received_callback */,
+                         barrier_callback /* data_completed_callback */);
+  }
+}
+
+void EngineConsumerConversationAPI::MergeSuggestTopicsResults(
+    GetSuggestedTopicsCallback callback,
+    std::vector<GenerationResult> results) {
+  // Merge the result and send another request to dedupe topics.
+  DedupeTopics(GetStrArrFromTabOrganizationResponses(results),
```

Comment:
Currently it's always called, due to prompt used for getting suggested topics and dedupe topics is a bit different and the latter one includes appending the emoji we need, this could probably be optimized in a follow-up.